### PR TITLE
test(downloader): add downloader test case

### PR DESCRIPTION
**Pull request by summingyu**
## Description
1. improve coverage
2. Optimize return time in case of errors

## Current Behavior
image

image

## New Behavior (screenshots if needed)

image

image

### CODE REVIEW

File: pkg/util/downloader/downloader.go [hunk 1]
```diff
@@ -14,11 +14,6 @@ import (
 func Download(url, filename, targetDir string) (int64, error) {
 	log.Debugf("Target dir: %s.", targetDir)
 	log.Debugf("URL: %s.", url)
-	resp, err := http.Get(url)
-	if err != nil {
-		return 0, err
-	}
-
 	var fileName = filename
```
Comment (IronCore864) [inline pkg/util/downloader/downloader.go:17]: It seems this isn't necessary.

How about:

```go
if filename == "" {
    filename = filepath.Base(url)
}
```
?

File: pkg/util/downloader/downloader.go [hunk 2]
```diff
@@ -38,5 +32,10 @@ func Download(url, filename, targetDir string) (int64, error) {
 		return 0, err
 	}
 
+	resp, err := http.Get(url)
```
Comment (IronCore864) [inline pkg/util/downloader/downloader.go:35]: Very good adjustment. No need to download the file first if we can't even get the filename from the URL.

File: pkg/util/downloader/downloader_test.go [hunk 3]
```diff
@@ -13,6 +13,13 @@ var _ = Describe("Downloader", func() {
 	Context("Downloader test", func() {
 		var url = "https://github.com/devstream-io/devstream/releases/download/v0.0.1/argocdapp_0.0.1.so"
 		var targetDir = "tmp"
+
+		It("Should return err", func() {
```
Comment (IronCore864) [inline pkg/util/downloader/downloader_test.go:17]: We can improve the name, for example:

It should return error if filename is empty and can't be parsed from the url.

**Review by IronCore864 — CHANGES_REQUESTED** [review]

**Comment by summingyu** [comment]
I think you're right, I'll change it now

**Review by aFlyBird0 — APPROVED** [review]
This helps to improve the speed of testing and increase test coverage, great!

File: pkg/util/downloader/downloader_test.go (hunk 3 above)
Comment (aFlyBird0) [reply pkg/util/downloader/downloader_test.go:17]: Please pay attention here. @summingyu

File: pkg/util/downloader/downloader_test.go [hunk 4]
```diff
@@ -13,6 +13,19 @@ var _ = Describe("Downloader", func() {
...
+			size, err := downloader.Download(url, ".", targetDir)
+			Expect(err).To(HaveOccurred())
+			Expect(size).To(Equal(int64(0)))
+		})
+
+		It("when url and filename are empty, Shold return err", func() {
```
Comment (aFlyBird0) [inline pkg/util/downloader/downloader_test.go:23]: *When    *should

**Review by aFlyBird0 — CHANGES_REQUESTED** [review]

**Review by IronCore864 — APPROVED** [review]

**Comment by IronCore864** [comment]
Thanks @summingyu for contributing more tests, and refactoring the existing logic!

I updated some of the descriptions of the test cases directly to fix grammar issues; hope you don't mind.

@aFlyBird0's comment is also fixed. Merging now. Due to duplicated commit messages, I'm squashing instead of merging. @summingyu please pay closer attention in the future. All in all, well done!

**Pull request merged**